### PR TITLE
Array-backed generic tuple (#1015)

### PR DESCRIPTION
**Pull request by toxxmeister**
Implementation of https://github.com/arrow-kt/arrow/issues/1015

### CODE REVIEW

**Comment by pakoito** [comment]
I sincerely hope you used codegen for this. I've done it manually before and it was a PITA with just one file, I don't want to imagine 10 of them :D

File: modules/core/arrow-generic/src/main/kotlin/arrow/generic/tup/TupFocus.kt [hunk 1]
```diff
@@ -0,0 +1,170 @@
+@file:Suppress("UNCHECKED_CAST")
+
+package arrow.generic.tup
+
+fun @receiver:Suppress("unused") Tup0Plus.focus0(): Tup0 = Tup()
```
Comment (pakoito) [inline modules/core/arrow-generic/src/main/kotlin/arrow/generic/tup/TupFocus.kt]: TIL

**Comment by toxxmeister** [comment]
@pakoito Yes, I only had to do some reformatting in the end, and even that was painful due to its sheer size.
By the way, do you include your codegen files anywhere in the repository? It might be useful to have it somewhere for future refactoring.

**Comment by toxxmeister** [comment]
@pakoito 
I'll add the script tomorrow, it needs some cleaning up first.

You suggest using `toTup` for `A.with(b: B)` and `to` for the rest, if I understood you right?
Ideally, It'd be great if we could find a word which makes sense semantically, is short, and can be used for all arities (unlike `toTup`/`to`).
Of course, if we don't find anything (which I suspect we won't, since naming things in FP is usually harder than writing the code itself), `toTup` will work.

Another question is whether we should shy away from using `with` here since it's completely unambiguous from Kotlin's `with(T, T.() -> R)`.

**Comment by pakoito** [comment]
I'd rather stay unambiguous, yeah. We get enough pushback for coming up with new things, overloading the existing ones would get even more hahaha

Can do gitter for a naming session :D

**Comment by raulraja** [comment]
I'm concerned about having multiple TupleN representations in Arrow and the confusion that may create for people getting started or even just realizing they just wrote some code for an incompatible type. We need to understand how much bytecode this increases the artifact because if it's similar to the TupleN impl in core and this is more performant and does not expose am unsafe API to end users this should probably live in `core` instead of `generic` replacing the existing TupleN implementation. Tuples are in core becuse we need them in certain type classes like `Applicative`,

**Comment by JorgeCastilloPrz** [comment]
In an effort to clean the PR list we've decided to close all the ones that are older than 2 months and consider them obsolete except for very specific circumstances that we will evaluate per PR. In this case all this should be generated and will be highly influenced by Arrow meta. That, on top of the time this has been on standby requires us to declare it obsolete for the time being. Note this can always be brought back to life if we have it as a priority in our roadmap at any point in time. Thank you for all the effort since it was a really good job! 🙏

**Comment by raulraja** [comment]
@toxxmeister id you are still interested in exploring this with arrow meta type proofs it will be possible to encode a lot of the proposed here in a few weeks when the design is finished and published

**Pull request closed without merging**